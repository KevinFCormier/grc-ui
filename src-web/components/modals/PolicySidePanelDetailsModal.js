/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { updateModal } from '../../actions/common'
import {connect} from 'react-redux'
import { withRouter } from 'react-router-dom'
import resources from '../../../lib/shared/resources'
import _ from 'lodash'
import {AllPoliciesInCluster, AllClustersInPolicy} from '../../../lib/client/queries'
import { Query } from 'react-apollo'
import { Modal, Loading, InlineNotification } from 'carbon-components-react'
import ResourceOverviewModule from '../modules/SubResourceListModule'
import getResourceDefinitions from '../../definitions'
import msgs from '../../../nls/platform.properties'
import queryString from 'query-string'
import {GRC_SIDE_PANEL_REFRESH_INTERVAL_COOKIE} from '../../../lib/shared/constants'
import {getPollInterval} from '../../components/common/RefreshTimeSelect'

resources(() => {
  require('../../../scss/side-panel-modal.scss')
})

function getHeader(data, locale) {
  const kind = _.get(data, 'kind', '')
  let header = '', descr = '', percent = 0, violation = '0/0', cluster = undefined, policy = undefined, hubNamespace = undefined
  if (kind === 'HCMPolicyPolicy') {
    header = _.get(data, 'raw.metadata.name', '')
    descr = _.get(data, 'raw.metadata.description', '')
    violation = _.get(data, 'clusterCompliant', '0/0')
    hubNamespace = _.get(data, 'namespace')
    const [vioNum, totalNum] = violation.split('/')
    if (!isNaN(vioNum) && !isNaN(totalNum)) {
      percent = +vioNum / +totalNum
    }
    violation += ' ' + msgs.get('overview.top.informations.clusters', locale)
    policy = data.name
  } else if (kind === 'HCMPolicyCluster') {
    header = _.get(data, 'cluster', '')
    descr = _.get(data, '', '')
    violation = _.get(data, 'violation', '0/0')
    const [vioNum, totalNum] = violation.split('/')
    if (!isNaN(vioNum) && !isNaN(totalNum)) {
      percent = +vioNum / +totalNum
    }
    violation += ' ' + msgs.get('overview.top.informations.policies', locale)
    cluster = data.cluster
  }
  return {header, descr, percent, violation, cluster, policy, hubNamespace}
}

class PolicySidePanelDetailsModal extends React.PureComponent {

  handleModalClose = () => {
    const { updateModal, location, history } = this.props
    updateModal()

    //update url and refresh policy page after closing sidepanel
    //text search input filter will not be removed, which is controled by itself
    const paraURL = {}
    let op = ''
    const curentURL = queryString.parse(location.search)
    if(curentURL.index) {
      paraURL.index = curentURL.index
      op = '?'
    }
    if(curentURL.filters) {
      paraURL.filters = curentURL.filters
      op = '?'
    }
    history.push(`${location.pathname}${op}${queryString.stringify(paraURL)}`)
  }

  render(){
    const { title, data, resourceType, locale } = this.props
    const { header, descr, percent, violation, cluster, policy, hubNamespace } = getHeader(data, locale)
    const inapplicable = msgs.get('table.actions.inapplicable', locale)
    const pollInterval = getPollInterval(GRC_SIDE_PANEL_REFRESH_INTERVAL_COOKIE, 'sidePanel')
    return (
      <div className={'side-panel-modal'}>
        <Modal
          passiveModal
          open={this.props.open}
          onRequestClose={this.handleModalClose}
          modalHeading={title}
          primaryButtonDisabled={true}
        >
          <div>
            <div className={'bx--modal-header special-header'}>
              {header}
              <div className={'bx--modal-content-header-content'}>
                {descr}
              </div>
              <div className={'bx--modal-content-header-main'}>
                <ProcessBar percent={percent} />
                <div className={'bx-modal-content-sub-header'}>{`${violation}`}</div>
              </div>
            </div>
            <div className={'bx--modal-content-body'}>
              <Query query={cluster ? AllPoliciesInCluster :  AllClustersInPolicy}
                variables={cluster ? {cluster} : {policy:policy, hubNamespace:hubNamespace}} pollInterval={pollInterval}>
                {( {data, loading, error} ) => {
                  if (loading) {
                    // spinner if loading
                    return <Loading withOverlay={false} />
                  }
                  if (error) {
                    // return inline error
                    return (
                      <InlineNotification key={`inline-notification-${error}`}
                        kind='error' title=''
                        subtitle={error.message}
                        iconDescription={'inline-error'} />
                    )
                  }
                  const { items } = data
                  if (cluster) {
                    return (
                      <ClustersTable items={items} type={resourceType} inapplicable={inapplicable} />
                    )
                  } else {
                    return (
                      <PoliciesTable items={items} type={resourceType} inapplicable={inapplicable} />
                    )
                  }
                }
                }
              </Query>
            </div>
          </div>
        </Modal>
      </div>
    )
  }
}

export const ClustersTable = ({items, type, inapplicable}) => {
  items = items.map((policy, index) => {
    let violatedNum = 0
    const spec = _.get(policy, 'raw.spec', '')
    const objectTemplates = _.get(spec, 'object-templates', [])
    const roleTemplates = _.get(spec, 'role-templates', [])
    const policyTemplates = _.get(spec, 'policy-templates', [])
    const id = _.get(policy, 'metadata.name', 'policy'+index)
    for (const template of objectTemplates) {
      if (_.get(template, 'status.Compliant','').toLowerCase() !== 'compliant') {
        violatedNum += 1
      }
    }
    for (const template of roleTemplates) {
      if (_.get(template, 'status.Compliant','').toLowerCase() !== 'compliant') {
        violatedNum += 1
      }
    }
    for (const template of policyTemplates) {
      if (_.get(template, 'status.Compliant','').toLowerCase() !== 'compliant') {
        violatedNum += 1
      }
    }

    if(violatedNum > 0) {
      const objectStatus = objectTemplates.map(item => {
        if (_.get(item, 'status.Compliant','').toLowerCase() !== 'compliant') {
          return {
            id: _.get(item, 'objectDefinition.metadata.name'),
            cells: [_.get(item, 'objectDefinition.metadata.name'), _.get(item, 'status.conditions[0].message', '-'), _.get(item, 'status.conditions[0].reason', '-')]
          }}}
      )
      const roleStatus = roleTemplates.map(item => {
        if (_.get(item, 'status.Compliant','').toLowerCase() !== 'compliant') {
          return {
            id: _.get(item, 'metadata.name'),
            cells: [_.get(item, 'metadata.name'), _.get(item, 'status.conditions[0].message', '-'), _.get(item, 'status.conditions[0].reason', '-')]
          }}}
      )
      const policyStatus = policyTemplates.map(item => {
        if (_.get(item, 'status.Compliant','').toLowerCase() !== 'compliant') {
          return {
            id: _.get(item, 'objectDefinition.metadata.name'),
            cells: [_.get(item, 'objectDefinition.metadata.name'), _.get(item, 'status.conditions[0].message', '-'), _.get(item, 'status.conditions[0].reason', '-')]
          }}}
      )
      //add id and remove null/undefined
      const subItems = _.without([id, ...objectStatus, ...roleStatus, ...policyStatus], undefined, null)
      return {...policy, id, violatedNum, subItems}
    }
    else {
      const subItems = [{ id: 'inapplicable'+index, cells: [inapplicable, inapplicable+' ', inapplicable+'  '] }]
      return {...policy, id, violatedNum, subItems}
    }
  })

  const staticResourceData = getResourceDefinitions(type)
  return (
    <div className='overview-content'>
      <ResourceOverviewModule
        staticResourceData={staticResourceData.clusterViolatedSidePanel}
        items={items}
        listSubItems={true}
      />
    </div>
  )
}

export const PoliciesTable = ({items, type, inapplicable}) => {
  items = items.map((cluster, index) => {
    const policy = _.get(cluster, 'policy', [])
    const violatedNum = _.get(cluster, 'violated', 0)
    const id = _.get(cluster, 'metadata.name', 'cluster'+index)

    if(violatedNum > 0) {
      const spec = _.get(policy, 'spec', '')
      const objectTemplates = _.get(spec, 'object-templates', [])
      const roleTemplates = _.get(spec, 'role-templates', [])
      const policyTemplates = _.get(spec, 'policy-templates', [])
      const subItems = _.without([
        id,
        ...objectTemplates.map(item => {
          if (_.get(item, 'status.Compliant','').toLowerCase() !== 'compliant') {
            return {
              id: _.get(item, 'objectDefinition.metadata.name'),
              cells: [_.get(item, 'objectDefinition.metadata.name'), _.get(item, 'status.conditions[0].message', '-'), _.get(item, 'status.conditions[0].reason', '-')]
            }}}
        ),
        ...roleTemplates.map(item => {
          if (_.get(item, 'status.Compliant','').toLowerCase() !== 'compliant') {
            return {
              id: _.get(item, 'objectDefinition.metadata.name'),
              cells: [_.get(item, 'objectDefinition.metadata.name'), _.get(item, 'status.conditions[0].message', '-'), _.get(item, 'status.conditions[0].reason', '-')]
            }}}
        ),
        ...policyTemplates.map(item => {
          if (_.get(item, 'status.Compliant','').toLowerCase() !== 'compliant') {
            return {
              id: _.get(item, 'objectDefinition.metadata.name'),
              cells: [_.get(item, 'objectDefinition.metadata.name'), _.get(item, 'status.conditions[0].message', '-'), _.get(item, 'status.conditions[0].reason', '-')]
            }}}
        )
      ], undefined, null)

      return {...cluster, id, subItems}
    }
    else {
      const subItems = [{ id: 'inapplicable'+index, cells: [inapplicable, inapplicable+' ', inapplicable+'  '] }]
      return {...cluster, id, subItems}
    }
  })

  const staticResourceData = getResourceDefinitions(type)

  //Link columns with fixed title name and target direction
  const linkFixedName = {2 : {fixedName:'table.actions.launch.cluster', urlTarget:'_blank'}}
  return (
    <div className='overview-content'>
      <ResourceOverviewModule
        staticResourceData={staticResourceData.policyViolatedSidePanel}
        items={items}
        listSubItems={true}
        linkFixedName={linkFixedName}
      />
    </div>
  )
}

const ProcessBar = ({percent}) => {
  const width = Math.min(window.innerWidth * 0.3, window.innerHeight * 0.4), height = 12, label = 'label', r = 6, w = width * percent
  return (
    <svg width={width} height={height} aria-label={label}>
      <rect className={'bx--modal-side-panel-process-bar-body'} width={width} height={height} rx={r} ry={r} />
      <rect className={'bx--modal-side-panel-process-bar-fill'} width={w} height={height} rx={r} ry={r} />
    </svg>
  )
}

ClustersTable.propTypes = {
  inapplicable: PropTypes.string,
  items: PropTypes.array,
  type: PropTypes.object,
}

PoliciesTable.propTypes = {
  inapplicable: PropTypes.string,
  items: PropTypes.array,
  type: PropTypes.object,
}

ProcessBar.propTypes = {
  percent: PropTypes.number,
}

PolicySidePanelDetailsModal.propTypes = {
  data: PropTypes.object,
  history: PropTypes.object.isRequired,
  locale: PropTypes.string,
  location: PropTypes.object,
  open: PropTypes.bool,
  resourceType: PropTypes.object,
  title: PropTypes.string,
  updateModal: PropTypes.func
}

const mapStateToProps = state =>  {
  return state.modal
}

const mapDispatchToProps = (dispatch) => {
  return {
    updateModal: () => dispatch(updateModal({ open: false }))
  }
}

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(PolicySidePanelDetailsModal))
