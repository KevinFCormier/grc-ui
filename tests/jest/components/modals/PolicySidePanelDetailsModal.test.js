/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import { shallow } from 'enzyme'
import { sidePanelPolicies, sidePanelResourceTypePolicies, sidePanelClusters, sidePanelResourceTypeClusters } from './ModalsTestingData'
import PolicySidePanelDetailsModal, { PoliciesTable, ClustersTable } from '../../../../src-web/components/modals/PolicySidePanelDetailsModal'
import renderer from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'

describe('PoliciesView view', () => {
  const updateModal = jest.fn()
  it('renders expand as expected', () => {
    const component = shallow(<PolicySidePanelDetailsModal
      title = {''}
      data = {sidePanelPolicies}
      resourceType = {sidePanelResourceTypePolicies}
      locale = {'en-US'}
      open = {true}
      updateModal = {updateModal}
    />)
    expect(component).toMatchSnapshot()
  })
})

describe('ClustersView view', () => {
  const updateModal = jest.fn()
  it('renders expand as expected', () => {
    const component = shallow(
      <PolicySidePanelDetailsModal
        title = {''}
        data = {sidePanelClusters}
        resourceType = {sidePanelResourceTypeClusters}
        locale = {'en-US'}
        open = {true}
        updateModal = {updateModal}
      />
    )
    expect(component).toMatchSnapshot()
  })
})

describe('PoliciesTable 1', () => {
  it('renders expand as expected', () => {
    const component = renderer.create(
      <BrowserRouter>
        <PoliciesTable
          items = {sidePanelPolicies}
          type = {sidePanelResourceTypePolicies}
          inapplicable = {'N/A'}
        />
      </BrowserRouter>
    )
    expect(component).toMatchSnapshot()
  })
})

describe('PoliciesTable 2', () => {
  it('renders expand as expected', () => {
    const component = renderer.create(
      <BrowserRouter>
        <PoliciesTable
          items = {[]}
          type = {sidePanelResourceTypePolicies}
          inapplicable = {'N/A'}
        />
      </BrowserRouter>
    )
    expect(component).toMatchSnapshot()
  })
})

describe('ClustersTable 1', () => {
  it('renders expand as expected', () => {
    const component = renderer.create(
      <BrowserRouter>
        <ClustersTable
          items = {sidePanelClusters}
          type = {sidePanelResourceTypeClusters}
          inapplicable = {'N/A'}
        />
      </BrowserRouter>
    )
    expect(component).toMatchSnapshot()
  })
})

describe('ClustersTable 2', () => {
  it('renders expand as expected', () => {
    const component = renderer.create(
      <BrowserRouter>
        <ClustersTable
          items = {[]}
          type = {sidePanelResourceTypeClusters}
          inapplicable = {'N/A'}
        />
      </BrowserRouter>
    )
    expect(component).toMatchSnapshot()
  })
})
