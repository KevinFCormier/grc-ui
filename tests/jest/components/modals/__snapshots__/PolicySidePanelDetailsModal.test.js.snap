// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolicySidePanelDetailsModal ClustersView view renders expand as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withRouter(Connect(PolicySidePanelDetailsModal))
    data={
      Object {
        "cluster": "cluster1",
        "custom": Object {
          "nonCompliant": Object {
            "_owner": null,
            "_store": Object {},
            "key": null,
            "props": Object {
              "maxCharacters": 35,
              "text": "policy-ma",
            },
            "ref": null,
          },
        },
        "kind": "HCMPolicyCluster",
        "namespace": "mcm",
        "nonCompliant": Array [
          "policy-ma",
        ],
        "violation": "1/4",
      }
    }
    locale="en-US"
    open={true}
    resourceType={
      Object {
        "list": "HCMPolicyClusterList",
        "name": "HCMPolicyCluster",
      }
    }
    title=""
    updateModal={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "render": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "render": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`PolicySidePanelDetailsModal PoliciesView view renders expand as expected 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withRouter(Connect(PolicySidePanelDetailsModal))
    data={
      Object {
        "__typename": "Compliance",
        "clusterCompliant": "1/2",
        "custom": Object {
          "metadata.annotations[\\"policy.mcm.ibm.com/categories\\"]": "System And Information Integrity",
          "metadata.annotations[\\"policy.mcm.ibm.com/controls\\"]": "Mutation Advisor",
          "metadata.annotations[\\"policy.mcm.ibm.com/standards\\"]": "NIST",
          "metadata.name": Object {
            "_owner": null,
            "_store": Object {},
            "key": null,
            "props": Object {
              "children": "policy-ma",
              "replace": false,
              "to": "/multicloud/policies/all/mcm/policy-ma",
            },
            "ref": null,
          },
        },
        "kind": "HCMPolicyPolicy",
        "metadata": Object {
          "__typename": "Metadata",
          "annotations": Object {
            "kubectl.kubernetes.io/last-applied-configuration": "{\\"apiVersion\\":\\"policy.mcm.ibm.com/v1alpha1\\",\\"kind\\":\\"Policy\\",\\"metadata\\":{\\"annotations\\":{\\"policy.mcm.ibm.com/categories\\":\\"SystemAndInformationIntegrity\\",\\"policy.mcm.ibm.com/controls\\":\\"MutationAdvisor\\",\\"policy.mcm.ibm.com/standards\\":\\"NIST\\"},\\"name\\":\\"policy-ma\\",\\"namespace\\":\\"mcm\\"},\\"spec\\":{\\"policy-templates\\":[{\\"objectDefinition\\":{\\"apiVersion\\":\\"policies.ibm.com/v1alpha1\\",\\"kind\\":\\"MutationPolicy\\",\\"metadata\\":{\\"label\\":{\\"category\\":\\"System-Integrity\\"},\\"name\\":\\"mutation-policy-example\\"},\\"spec\\":{\\"conditions\\":{\\"ownership\\":[\\"ReplicaSet\\",\\"Deployment\\",\\"DeamonSet\\",\\"ReplicationController\\"]},\\"namespaceSelector\\":{\\"exclude\\":[\\"kube-system\\"],\\"include\\":[\\"default\\",\\"kube-*\\"]},\\"remediationAction\\":\\"enforce\\"}}}],\\"remediationAction\\":\\"enforce\\"}}
",
            "policy.mcm.ibm.com/categories": "SystemAndInformationIntegrity",
            "policy.mcm.ibm.com/controls": "MutationAdvisor",
            "policy.mcm.ibm.com/standards": "NIST",
            "seed-generation": "1",
          },
          "name": "policy-ma",
          "namespace": "mcm",
          "resourceVersion": "1347579",
          "selfLink": "/apis/policy.mcm.ibm.com/v1alpha1/namespaces/mcm/policies/policy-ma",
        },
        "name": "policy-ma",
        "namespace": "mcm",
        "placementBindings": Array [
          Object {
            "__typename": "PlacementBinding",
            "metadata": Object {
              "__typename": "Metadata",
              "name": "binding-ma",
              "selfLink": "/apis/mcm.ibm.com/v1alpha1/namespaces/mcm/placementbindings/binding-ma",
            },
          },
        ],
        "placementPolicies": Array [
          Object {
            "__typename": "PlacementPolicy",
            "metadata": Object {
              "__typename": "Metadata",
              "name": "placement-ma",
              "selfLink": "/apis/mcm.ibm.com/v1alpha1/namespaces/mcm/placementpolicies/placement-ma",
            },
          },
        ],
        "policyCompliant": "1/2",
        "raw": Object {
          "apiVersion": "policy.mcm.ibm.com/v1alpha1",
          "kind": "Policy",
          "metadata": Object {
            "annotations": Object {
              "kubectl.kubernetes.io/last-applied-configuration": "{\\"apiVersion\\":\\"policy.mcm.ibm.com/v1alpha1\\",\\"kind\\":\\"Policy\\",\\"metadata\\":{\\"annotations\\":{\\"policy.mcm.ibm.com/categories\\":\\"SystemAndInformationIntegrity\\",\\"policy.mcm.ibm.com/controls\\":\\"MutationAdvisor\\",\\"policy.mcm.ibm.com/standards\\":\\"NIST\\"},\\"name\\":\\"policy-ma\\",\\"namespace\\":\\"mcm\\"},\\"spec\\":{\\"policy-templates\\":[{\\"objectDefinition\\":{\\"apiVersion\\":\\"policies.ibm.com/v1alpha1\\",\\"kind\\":\\"MutationPolicy\\",\\"metadata\\":{\\"label\\":{\\"category\\":\\"System-Integrity\\"},\\"name\\":\\"mutation-policy-example\\"},\\"spec\\":{\\"conditions\\":{\\"ownership\\":[\\"ReplicaSet\\",\\"Deployment\\",\\"DeamonSet\\",\\"ReplicationController\\"]},\\"namespaceSelector\\":{\\"exclude\\":[\\"kube-system\\"],\\"include\\":[\\"default\\",\\"kube-*\\"]},\\"remediationAction\\":\\"enforce\\"}}}],\\"remediationAction\\":\\"enforce\\"}}
",
              "policy.mcm.ibm.com/categories": "SystemAndInformationIntegrity",
              "policy.mcm.ibm.com/controls": "MutationAdvisor",
              "policy.mcm.ibm.com/standards": "NIST",
              "seed-generation": "1",
            },
            "creationTimestamp": "2019-07-16T13:18:48Z",
            "finalizers": Array [
              "propagator.finalizer.mcm.ibm.com",
            ],
            "generation": 45,
            "name": "policy-ma",
            "namespace": "mcm",
            "resourceVersion": "1347579",
            "selfLink": "/apis/policy.mcm.ibm.com/v1alpha1/namespaces/mcm/policies/policy-ma",
            "uid": "3f218d45-a7cc-11e9-8a41-005056a061f1",
          },
          "spec": Object {
            "namespaces": Object {},
            "policy-templates": Array [
              Object {
                "objectDefinition": Object {
                  "apiVersion": "policies.ibm.com/v1alpha1",
                  "kind": "MutationPolicy",
                  "metadata": Object {
                    "label": Object {
                      "category": "System-Integrity",
                    },
                    "name": "mutation-policy-example",
                  },
                  "spec": Object {
                    "conditions": Object {
                      "ownership": Array [
                        "ReplicaSet",
                        "Deployment",
                        "DeamonSet",
                        "ReplicationController",
                      ],
                    },
                    "namespaceSelector": Object {
                      "exclude": Array [
                        "kube-system",
                      ],
                      "include": Array [
                        "default",
                        "kube-*",
                      ],
                    },
                    "remediationAction": "enforce",
                  },
                },
                "status": Object {
                  "Validity": Object {},
                },
              },
            ],
            "remediationAction": "enforce",
          },
          "status": Object {
            "placementBindings": Array [
              "binding-ma",
            ],
            "placementPolicies": Array [
              "placement-ma",
            ],
            "status": Object {
              "cluster1": Object {
                "aggregatePoliciesStatus": Object {
                  "policy-ma": Object {
                    "compliant": "NonCompliant",
                  },
                },
                "clustername": "cluster1",
                "compliant": "NonCompliant",
              },
              "clusterhub": Object {
                "aggregatePoliciesStatus": Object {
                  "policy-ma": Object {
                    "compliant": "Compliant",
                  },
                },
                "clustername": "clusterhub",
                "compliant": "Compliant",
              },
            },
          },
        },
        "remediation": "enforce",
        "subItems": Array [
          Object {
            "items": Array [
              "binding-ma",
            ],
            "name": "policy.pb",
          },
          Object {
            "items": Array [
              "placement-ma",
            ],
            "name": "policy.pp",
          },
        ],
      }
    }
    locale="en-US"
    open={true}
    resourceType={
      Object {
        "list": "HCMPolicyPolicyList",
        "name": "HCMPolicyPolicy",
      }
    }
    title=""
    updateModal={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "render": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "render": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
